Fixed #28690 django.utils.http.parse_http_date two digit year check is incorrect

Baguage opened the pull request:
https://code.djangoproject.com/ticket/28690

claudep commented on tests/utils_tests/test_http.py:
```diff
@@ -223,6 +223,21 @@ def test_parsing_rfc850(self):
+            ('Monday, 15-Aug-51 08:49:37 GMT', datetime(2051, 8, 15, 8, 49, 37)),
+            ('Monday, 15-Aug-70 00:00:00 GMT', datetime(1970, 8, 15, 0, 00, 00)),
+            ('Monday, 15-Aug-71 08:49:37 GMT', datetime(1971, 8, 15, 8, 49, 37)),
+            ('Monday, 15-Aug-95 08:49:37 GMT', datetime(1995, 8, 15, 8, 49, 37)),
```
Why so much lines? Is each line testing a different code path?

Baguage replied:
Updated, only two tests for two different branches are left

atombrella replied:
I don't think the number of tests is an issue, but since the check depends on a variable, the test should also reflect this, or it'll break at some point.

atombrella commented on django/utils/http.py:
```diff
@@ -144,7 +144,14 @@ def parse_http_date(date):
         year = int(m.group('year'))
         if year < 100:
-            if year < 70:
+            # RFC 850 does not mention this, but in RFC 7231 (and there's something similar in RFC 2822),
```
I don't think repeating the comment from the ticket makes sense. RFC 7231 section 7.1.1.1 contains the same information, and is already referenced in the doc-string.

Baguage replied:
I removed the comments.
Not 100% sure if I understand your note about #23919 - that ticket doesn't seem to mention emails.Util. Could you elaborate? Thanks!

timgraham commented on tests/utils_tests/test_http.py:
```diff
@@ -223,6 +223,16 @@ def test_parsing_rfc850(self):
+        )
+        for rfc850str, rfc850date in dates:
+            with self.subTest(string=rfc850str, date=rfc850date):
+                parsed = http.parse_http_date(rfc850str)
```
Please rebase your branch on master and remove `http.` after 0e212a705e6b2e49a246b16286036c40ec2ac4f8.

timgraham commented on tests/utils_tests/test_http.py:
```diff
@@ -223,6 +223,16 @@ def test_parsing_rfc850(self):
+            ('Monday, 15-Aug-70 00:00:00 GMT', datetime(1970, 8, 15, 0, 00, 00)),
+        )
+        for rfc850str, rfc850date in dates:
+            with self.subTest(string=rfc850str, date=rfc850date):
```
`date=rfc850date` isn't needed in the `subTest()` -- since will appear if the assertion fails.

Baguage commented:
@timgraham I added regression test that fails with old code (test_parsing_rfc850_year_69), updated commit message to hopefully follow the guidelines, and added additional comments about the change. Squashed commits as well.

timgraham commented:
Only regression tests should be added in the fix. If you want to add other tests in a separate commit that precedes the fix, that's fine. I would use a loop and `self.subTest()` rather than having three test methods that differ in only their input values. Also, I think the test dates need to be generated relative to `today()`, otherwise the expected value will change as time goes on, won't it?

Baguage commented:
> Only regression tests should be added in the fix. If you want to add other tests in a separate commit that precedes the fix, that's fine. I would use a loop and self.subTest() rather than having three test methods that differ in only their input values. 

Done!

> the test dates need to be generated relative to today(), otherwise the expected value will change as time goes on, won't it?

Indeed. I thought the tests I wrote will work in future, but after playing with the dates a bit I realized I was wrong. Corrected!

Is is possible to re-run the windows test in Jenkins? It failed with a strange error, and I'm not sure what I can do to fix it:

     hudson.plugins.git.GitException: Command "C:\Program Files\Git\bin\git.exe checkout -f 56f53b90ebdbb249fa7d9cc310012377a10d1af7" returned status code 128:

timgraham commented on tests/utils_tests/test_http.py:
```diff
@@ -268,6 +268,18 @@ def test_parsing_rfc850(self):
+            ('Wednesday, 31-Dec-%s 18:49:37 GMT' % (year + 51), datetime(1900 + year + 51, 12, 31, 18, 49, 37)),
+        )
+        for rfc850str, rfc850date in dates:
+            with self.subTest(string=rfc850str, date=rfc850date):
```
Don't include `date=rfc850date` in the `subTest()`. That string will already be displayed if the assertion fails.

timgraham commented on tests/utils_tests/test_http.py:
```diff
@@ -268,6 +268,18 @@ def test_parsing_rfc850(self):
+    def test_parsing_rfc850_year_50(self):
+        year = datetime.now().year % 100
+        dates = (
+            ('Monday, 31-Dec-18 08:49:37 GMT', datetime(2018, 12, 31, 8, 49, 37)),
```
At what point will this be interpreted as 2118? Far enough into the future that it doesn't matter much? Might want to include a comment about that.

Baguage commented:
Thank you for bearing with me, Tim. I will put additional tests in a separate commit, fix subTests loop and note about when 31-Dec-18 will be interpreted as 2118 (around the year 2068, I believe - so this probably doesn't matter)

timgraham commented:
Have time to finish it?

felixxm commented:
Superseded by #11212.
